### ISSUE TITLE
Clarification on BAI distinct bins

### ISSUE THREAD

**Issue by innovate-invent** [samtools/hts-specs#284]
I am assuming the 'bin' field in the format specification is there to identify the bin. Why is it there?
Are bins allowed to be missing or out of order?
If so, is there a specification on what bin numbers to use and their order?
I understand that the bin numbers correspond to a bin size but what determines the bin within those ranges?

Thanks

**Comment by jmarshall** [comment]
§5.2 on BAI indices is one of the older parts of the spec and it is agreed that there are a number of details that are not spelt out. At all.

The number of reads in each of the bins of course depends on the reads in the particular BAM file, and it is usually the case that some bins will have no reads in them. Consider for example a BAM file from a panel-based experiment that might aim to have reads only in say chr1:1000000-1200000: practically all the bins representing intervals outside that region will be empty.

Empty bins are allowed to be omitted from the BAI/CSI/etc index; in fact, it's probably encouraged.

The specification does not state an ordering for the bins within the “list of distinct bins” for a reference sequence. And in fact there exist implementations (notably samtools/HTSlib) that write them out in a semi-random order, as they simply dump out their in-memory hash table. (Yes, I was surprised the first time I noticed that!) So you can't assume that bins are in any particular order.

The table in the specification says “List of **distinct bins**”, which is an overly-brief way of saying that (for each particular reference sequence) the bin numbers in the list must be distinct. It wouldn't make sense to write out a BAI/CSI index with the same bin number appearing more than once (for a reference sequence), as the point of the index is to collect together all the chunks for each bin.

I suppose for tools reading BAI/CSI index files it might be rational to concatenate the lists of chunks seen if a bin occurs more than once, but implementations aren't required to do that. The specification says that they will be distinct, and for example samtools/HTSlib returns an error if asked to read a BAI file with duplicate bins.

**Comment by jkbonfield** [comment]
I think it is implicit that given we have a "list of distinct bins" which has a field to indicate which bin number we are describing, that it is permitted to skip empty ones.  If that wasn't the case we'd not need to have a field indicating the bin number and instead would just have an array instead of a list so I don't really see a strong need for expanding on that.  The (lack of) ordering though is something that should be stated.

If you have recommended wording to make more of this explicit though please suggest it.

**Comment by innovate-invent** [comment]
I agree with @jmarshall on the wording, it leaves too much ambiguity. It isn't clear what makes the bin distinct or that the bin numbers specifically need to be unique.

> If that wasn't the case we'd not need to have a field indicating the bin number and instead would just have an array instead

That was my original thought but it wouldn't be the first aspect of this spec that is a little jenky for the sake of backwards compatibility. I would expect a specification to minimize ambiguity to maximize reproducibility.  
It would be good to have "for each particular reference sequence the bin numbers in the list must be distinct" somewhere in the description.
